### ISSUE TITLE
File descriptor leak in salt.client when using the Caller API

### ISSUE THREAD

**Issue by christoe** [saltstack/salt#39717]
### File descriptor leak in salt.client when using the Caller API

Looks very similar to the issue described in #27884 (which was closed without a comment), except that we are using a full setup rather than a local client.

### Steps to Reproduce Issue
```
from salt import client
client.Caller().cmd('state.highstate', test=True)
```

Every invocation of this results in more file descriptors being allocated. It seems as if some of them are garbage collected from time to time, but in our use case, they are allocated a lot faster than they are deallocated.

Excerpt from `ls la /proc/<pid>/fd`:
```
[...]
0 lrwx------. 1 root root 64 Feb 28 14:49 20 -> socket:[144901766]
0 lr-x------. 1 root root 64 Feb 28 14:49 21 -> pipe:[144902892]
0 l-wx------. 1 root root 64 Feb 28 14:49 22 -> pipe:[144902892]
0 lrwx------. 1 root root 64 Feb 28 14:49 23 -> anon_inode:[eventfd]
0 lrwx------. 1 root root 64 Feb 28 14:49 24 -> anon_inode:[eventfd]
0 lrwx------. 1 root root 64 Feb 28 14:49 25 -> anon_inode:[eventpoll]
0 lrwx------. 1 root root 64 Feb 28 14:49 26 -> anon_inode:[eventfd]
0 lrwx------. 1 root root 64 Feb 28 14:49 27 -> anon_inode:[eventpoll]
0 lrwx------. 1 root root 64 Feb 28 14:49 28 -> anon_inode:[eventfd]
[...]
```

### Versions Report
```
# salt --versions-report
Salt Version:
           Salt: 2016.11.2

Dependency Versions:
           cffi: 0.8.6
       cherrypy: 3.2.2
       dateutil: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.21.0
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
         pygit2: 0.21.4
         Python: 2.7.5 (default, Sep 15 2016, 22:37:39)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4

System Versions:
           dist: centos 7.2.1511 Core
        machine: x86_64
        release: 3.10.0-327.36.3.el7.x86_64
         system: Linux
        version: CentOS Linux 7.2.1511 Core
```

**Comment by gtmanfred** [comment]
The caller client is like salt-call.  It runs completely separate of the running salt minion, it generates its own connection to the salt master, and runs its own maintenance tasks.

Where exactly are you seeing these File descriptors get leaked?  Is it in a script that you use that constantly runs and does the caller client?  Are you constantly initializing the Caller client?

Is there any other information you can give me about how this is setup?

Thanks,
Daniel

**Comment by christoe** [comment]
We have a small script that exposes some Salt metrics on a built in web server. To collect these metrics we make calls using the caller client on a regular basis.

The file descriptors seem to leak on every call to `Caller().cmd()`. I've tried both reusing the same Caller client and reinitializing it on every call, but both seem to yield the same result. I've also tried to explicitly `del` the caller client, but that does not seem to have any effect either.

Can you give me any hints on where to continue looking as to provide you with more information?

**Comment by carlpett** [comment]
It seems this was fixed between 2016.11.2 and 2016.11.3. I wrote a small tester util, https://gist.github.com/carlpett/7bd0bf087658056a7e863b1d1fdd19a9. If I run this on 2016.11.2, every iteration creates _a lot_ of objects, including new AsyncReqMessageClientPool, ZMQStream, Socket and AsyncZeroMQReqChannel. After ~40 iterations, I have ~400 fds, and my VM runs out of memory. 

When I upgrade to 2016.11.3, this no longer happens. There aren't any obvious changes in the changelist, so it seems to have been "accidentally" fixed. It would be nice to determine what fixed it, so that it doesn't regress in a later release. I tried following the code path but quicky got lost, unfortunately.

**Comment by gtmanfred** [comment]
I will try to get around to doing a git bisect on this, but unfortunately, if it is not an issue in the newest release, we have to prioritize other issues.

Thanks,
Daniel

**Comment by carlpett** [comment]
@gtmanfred I'll see if I can get around to setting up a full-blown salt-dev environment. Currently I have it working well enough to do pull requests and unit tests, but actually running it seems to be a bit away still...

**Comment by gtmanfred** [comment]
Let me know if you can get it bisected.

Otherwise, I am going to put it on my schedule to try and do this next week.

**Comment by carlpett** [comment]
Ok, I think I got it worked out. This seems to be the one that fixes it: https://github.com/saltstack/salt/commit/0b18f3467817e27d3039fd27bb806ecb087f2dd3

Another thing to note is that it seems not all state modules will trigger the issue. This, for example, works even before that commit:

```yaml
my-state:
  cmd.run:
    - name: echo Hello
```

While this does not:

```yaml
my-state:
  service.running:
    - name: sshd
```

**Comment by christoe** [comment]
Saltstack 2016.11.3 does indeed solve this issue as @carlpett suggested.

**Comment by gtmanfred** [comment]
I have confirmed this as well, This was fixed by https://github.com/saltstack/salt/pull/38951 Which was meant to fix #33890

Thanks to everyone, I am going to close this issue now.
Daniel

**Comment by carlpett** [comment]
@gtmanfred Looking at the fixing PR, it is pretty unclear that this could have the sort of effect as reported here and in the other bug report. Would it be wise to put some cautious comment in that file that changing it can lead to non-obvious effects? Or maybe this is already clear to the people who would be editing it? 
Just not keen on a regression on this, but not sure how to best avoid it.
